### DISCUSSION TITLE
Fortran wrapper module for onnx API?

### DISCUSSION THREAD

**Discussion by jalvesz** [onnx/onnx#4871] [Q&A]
Hi, I was wondering if there is any activity that could help using onnx directly from Fortran code? This could be very useful for HPC applications using Fortran as the code base. I'm currently looking at how to create a minimal wrapper, any help/guidance here would be very much appreciated.
Thanks

**Comment by jalvesz** [comment]
Hi, yes I'm looking at ways to create the appropriate wrappers for Fortran from the C headers provided with iso_c_binding. F2PY is not an option, the objective is to work directly on Fortran not on python. I'm not sure what did you ment with CAF? I'm only familiar with CAF for CoArray Fortran, which is a framework for distributed parallel computing directly in the language.

**Reply by jalvesz** [reply]
I started looking at libraries such as Shroud https://github.com/LLNL/shroud or SWIFT-Fortran to create the wrappers.

**Comment by BriannaSensf** [comment]
Yes, there are options available to use ONNX directly from Fortran code. Here are some possible approaches:

Using C-Fortran interoperability: ONNX provides C bindings, and Fortran code can use C bindings through the interoperability features provided by the Fortran standard. You can create a Fortran wrapper around the ONNX C bindings to expose them to Fortran code.

Using a Fortran-to-C compiler: You can use a Fortran-to-C compiler, such as f2c or gfortran, to convert your Fortran code to C code, and then use the ONNX C bindings directly from the resulting C code.

Using a C-Fortran interface library: You can use a C-Fortran interface library, such as fwrap or cfortran, to create a Fortran wrapper around the ONNX C bindings. These libraries simplify the process of creating C-Fortran interfaces and provide a more Fortran-friendly API.

In general, the best approach will depend on your specific requirements and constraints. If you can provide more details about your use case, I may be able to provide more specific guidance.

**Reply by jalvesz** [reply]
Hi, thanks!!
So, as a minimal working example, I would like to compare the evaluation of a simple regression problem:
I have a problem that I can define with a function, lets say 'sigma = fun(x,y,z)'. For a simple case I do have the analytical expression, but for the complex cases this is defined with a huge database with a 3D grid on the 'x,y,z' values, thus the interest of creating an ANN.

I have already built a script using pytorch. I can either create synthetic data from the simple cases and train, or take the file data base for training a model like sigma=ANN(x,y,z).

In my Fortran code, I need to be able of calling this model element-wise, but also its derivatives with something like autograd;

...
use onnxruntime_fortran_api ! just imagining things here
integer :: n
real, allocatable :: x(:), y(:), z(:)
real, allocatable :: sigma(:), dsigma(:,:)
type( onnx_session_t ) :: model ! just imagining things here
...
allocate( sigma(n), dsigma(3,n) , x(n), y(n), z(n) )
...
call model%onnx_load( model_path=path_to_mymodel.onnx , model ) ! Guessing this should wrap "Ort::Session session(env, model_path, session_options);"
do i=1,n
sigma(i) = model%Run( x(i),y(i),z(i) )
dsigma(1,i) = model%autograd( sigma(i), x(i) , ... (?) ) ! following here a pytorch-like notation, not sure how it would be done with onnx
dsigma(2,i) = model%autograd( sigma(i), z(i) , ... (?) )
dsigma(3,i) = model%autograd( sigma(i), y(i) , ... (?) )
! OR
dsigma(1:3,i) = model%autograd( sigma(i), [x(i),y(i),z(i)] , ... (?) )
end do

(the actual loop is more complex, some pre and post data treatements are done before evaluating the model for each element.
And I won't be needing (for the moment) GPU execution, only CPU.

I know how to export the model using pytorch with torch.onnx.export(model, (x,y,z), "mymodel.onnx", input_names=['x','y','z'], output_names=['sigma'] )

Now the thing is

I'm not sure where to start in order to create the minimal wrapper for Fortran (just to load and infer) by looking at the C or C++ headers
less sure about the possibility of doing the differentiation such as described, a quick search in google did not gave me a clear clue on this one
I can put all of this into a "not-fully-working" minimal project and see if it is possible to get somewhere by iterating over the ideas ? Does it seem feasible?

**Comment by BrandenTaggartG** [comment]
There are several API wrappers available for Fortran, such as F2PY, SWIG, and CMake Fortran API. The choice of wrapper depends on the specific requirements of the project.
